File: probe_agent/lib/src/protocol.dart
import 'dart:convert';

/// JSON-RPC 2.0 request received from the probe CLI.
class ProbeRequest {
  final String jsonrpc;
  final int id;
  final String method;
  final Map<String, dynamic> params;

  const ProbeRequest({
    required this.jsonrpc,
    required this.id,
    required this.method,
    required this.params,
  });

  factory ProbeRequest.fromJson(Map<String, dynamic> json) {
    return ProbeRequest(
      jsonrpc: json['jsonrpc'] as String? ?? '2.0',
      id: (json['id'] as num).toInt(),
      method: json['method'] as String,
      params: (json['params'] as Map<String, dynamic>?) ?? {},
    );
  }

  static ProbeRequest? tryParse(String raw) {
    try {
      final map = jsonDecode(raw) as Map<String, dynamic>;
      if (!map.containsKey('id')) return null; // notification
      return ProbeRequest.fromJson(map);
    } catch (_) {
      return null;
    }
  }
}

/// JSON-RPC 2.0 response sent back to the probe CLI.
class ProbeResponse {
  final int id;
  final dynamic result;
  final ProbeError? error;

  const ProbeResponse.ok(this.id, this.result) : error = null;
  const ProbeResponse.err(this.id, this.error) : result = null;

  Map<String, dynamic> toJson() {
    if (error != null) {
      return {'jsonrpc': '2.0', 'id': id, 'error': error!.toJson()};
    }
    return {'jsonrpc': '2.0', 'id': id, 'result': result};
  }

  String encode() => jsonEncode(toJson());
}

/// JSON-RPC 2.0 error object.
class ProbeError {
  final int code;
  final String message;

  const ProbeError(this.code, this.message);

  // Standard codes
  static const int parseError     = -32700;
  static const int methodNotFound = -32601;
  static const int invalidParams  = -32602;
  static const int internalError  = -32603;
  static const int timeout        = -32000;
  static const int widgetNotFound = -32001;
  static const int assertFailed   = -32002;

  Map<String, dynamic> toJson() => {'code': code, 'message': message};
}

/// JSON-RPC 2.0 notification (no id) sent from agent to CLI.
class ProbeNotification {
  final String method;
  final dynamic params;

  const ProbeNotification(this.method, this.params);

  String encode() => jsonEncode({
        'jsonrpc': '2.0',
        'method': method,
        'params': params,
      });
}

/// JSON-RPC method name constants — mirrors Go probelink/protocol.go.
class ProbeMethods {
  ProbeMethods._();

  // Request methods (CLI → agent)
  static const ping = 'probe.ping';
  static const settled = 'probe.settled';
  static const open = 'probe.open';
  static const tap = 'probe.tap';
  static const doubleTap = 'probe.double_tap';
  static const longPress = 'probe.long_press';
  static const type_ = 'probe.type';
  static const clear = 'probe.clear';
  static const see = 'probe.see';
  static const wait = 'probe.wait';
  static const swipe = 'probe.swipe';
  static const scroll = 'probe.scroll';
  static const drag = 'probe.drag';
  static const deviceAction = 'probe.device_action';
  static const close = 'probe.close';
  static const screenshot = 'probe.screenshot';
  static const dumpTree = 'probe.dump_tree';
  static const saveLogs = 'probe.save_logs';
  static const runDart = 'probe.run_dart';
  static const mock = 'probe.mock';
  static const startRecording = 'probe.start_recording';
  static const stopRecording = 'probe.stop_recording';

  // Notification methods (agent → CLI)
  static const notifyRecordedEvent = 'probe.recorded_event';
  static const notifyExecDart = 'probe.exec_dart';
  static const notifyRestartApp = 'probe.restart_app';
}

File: internal/probelink/protocol.go
package probelink

import (
	"encoding/json"
	"fmt"
	"sync/atomic"
)

// JSON-RPC 2.0 message types used by ProbeLink.

var idCounter uint64

func nextID() uint64 {
	return atomic.AddUint64(&idCounter, 1)
}

// Request is a JSON-RPC 2.0 request.
type Request struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      uint64          `json:"id"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params,omitempty"`
}

// Response is a JSON-RPC 2.0 response.
type Response struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      uint64          `json:"id"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *RPCError       `json:"error,omitempty"`
}

// RPCError is a JSON-RPC 2.0 error object.
type RPCError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

func (e *RPCError) Error() string {
	return fmt.Sprintf("rpc error %d: %s", e.Code, e.Message)
}

// Notification is a JSON-RPC 2.0 notification (no ID).
type Notification struct {
	JSONRPC string          `json:"jsonrpc"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params,omitempty"`
}

// NewRequest creates a new JSON-RPC request with an auto-incremented ID.
func NewRequest(method string, params any) (*Request, error) {
	raw, err := json.Marshal(params)
	if err != nil {
		return nil, err
	}
	return &Request{
		JSONRPC: "2.0",
		ID:      nextID(),
		Method:  method,
		Params:  raw,
	}, nil
}

// ---- Command parameter types ----

// TapParams targets a widget by selector.
type TapParams struct {
	Selector SelectorParam `json:"selector"`
}

// TypeParams enters text into a widget.
type TypeParams struct {
	Selector SelectorParam `json:"selector"`
	Text     string        `json:"text"`
}

// SeeParams checks widget visibility.
type SeeParams struct {
	Selector SelectorParam `json:"selector"`
	Negated  bool          `json:"negated"`
	Count    int           `json:"count,omitempty"`
	Check    string        `json:"check,omitempty"`
	CheckVal string        `json:"check_val,omitempty"`
	Pattern  string        `json:"pattern,omitempty"`
}

// WaitParams controls wait behaviour.
type WaitParams struct {
	Kind     string  `json:"kind"`   // appears | disappears | page_load | network_idle | duration
	Target   string  `json:"target,omitempty"`
	Duration float64 `json:"duration,omitempty"` // seconds
	Timeout  float64 `json:"timeout,omitempty"`
}

// SwipeParams controls a swipe gesture.
type SwipeParams struct {
	Direction string        `json:"direction"` // up | down | left | right
	Selector  *SelectorParam `json:"selector,omitempty"`
}

// ScrollParams controls a scroll action.
type ScrollParams struct {
	Direction string        `json:"direction"`
	Selector  *SelectorParam `json:"selector,omitempty"`
}

// OpenParams opens the app or a named screen.
type OpenParams struct {
	Screen string `json:"screen,omitempty"` // empty = launch app
}

// SelectorParam describes a widget locator.
type SelectorParam struct {
	Kind      string `json:"kind"`               // text | id | type | ordinal | positional
	Text      string `json:"text,omitempty"`
	Ordinal   int    `json:"ordinal,omitempty"`
	Container string `json:"container,omitempty"`
}

// ScreenshotResult is returned by the take_screenshot command.
type ScreenshotResult struct {
	Path string `json:"path"` // absolute path on host machine
}

// MockParam registers an HTTP mock.
type MockParam struct {
	Method string `json:"method"`
	Path   string `json:"path"`
	Status int    `json:"status"`
	Body   string `json:"body,omitempty"`
}

// DartParam runs Dart code on the agent.
type DartParam struct {
	Code string `json:"code"`
}

// GenericResult is used when the response is a simple OK/error.
type GenericResult struct {
	OK      bool   `json:"ok"`
	Message string `json:"message,omitempty"`
}

// WidgetTreeResult contains the serialised widget tree.
type WidgetTreeResult struct {
	Tree string `json:"tree"`
}

// DeviceActionParams handles device-level actions.
type DeviceActionParams struct {
	Action string `json:"action"` // rotate_landscape | rotate_portrait | toggle_dark_mode | set_locale | ...
	Value  string `json:"value,omitempty"`
}

// ---- Method names ----

const (
	MethodOpen         = "probe.open"
	MethodTap          = "probe.tap"
	MethodType         = "probe.type"
	MethodSee          = "probe.see"
	MethodWait         = "probe.wait"
	MethodSwipe        = "probe.swipe"
	MethodScroll       = "probe.scroll"
	MethodLongPress    = "probe.long_press"
	MethodDoubleTap    = "probe.double_tap"
	MethodClear        = "probe.clear"
	MethodClose        = "probe.close"
	MethodDrag         = "probe.drag"
	MethodScreenshot   = "probe.screenshot"
	MethodDumpTree     = "probe.dump_tree"
	MethodRunDart      = "probe.run_dart"
	MethodMock         = "probe.mock"
	MethodDeviceAction = "probe.device_action"
	MethodPing           = "probe.ping"
	MethodSettled        = "probe.settled"   // wait for triple-signal sync
	MethodSaveLogs       = "probe.save_logs"
	MethodStartRecording = "probe.start_recording"
	MethodStopRecording  = "probe.stop_recording"

	// Notification methods (agent → CLI, no response expected)
	NotifyRecordedEvent = "probe.recorded_event"
	NotifyExecDart      = "probe.exec_dart"
	NotifyRestartApp     = "probe.restart_app"
)
